rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,522 @@
+* (-) This representation imposes a limitation on the evaluation order.
+  Since we can’t express the computation between two intermediate (accumulation)
+  results, we can not arbitrarily insert `init_value` and start the
+  computation at an arbitrary location. The following shows the restricted
+  evaluation order with the method.
+
+```python
++----------+       +--------+ +--------+ +--------+ +--------+
+|init_value|       |input[0]| |input[1]| |input[2]| |input[3]|
++----------+       +--------+ +--------+ +--------+ +--------+
+           \        /         /          /          /
+           +-------+         /          /          /
+           |compute|        /          /          /
+           +-------+       /          /          /
+                \         /          /          /
+                 +-------+          /          /
+                 |compute|         /          /
+                 +-------+        /          /
+                       \         /          /
+                        +-------+          /
+                        |compute|         /
+                        +-------+        /
+                              \         /
+                               +-------+
+                               |compute|
+                               +-------+
+```
+
+## Open Question
+
+### Should we restrict the proposal #1 to quantized types only?
+
+The above proposal #1 of introducing the additional functions is theoretically
+not limited to quantized `reduce` op, but also can be applied to `reduce` op with
+non-quantized types. For example,
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+  ^input_conversion(%arg0: tensor<bf16>):
+    %0 = "stablehlo.convert"(%arg0): (tensor<bf16>) -> (tensor<f32>)
+    "stablehlo.return"(%0) : (tensor<f32>) -> (tensor<f32>)
+  }, {
+  ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
+    %0 = "stablehlo.add"(%arg0, %arg1) : (tensor<f32>, tensor<f32>) ->
+    tensor<f32>
+    "stablehlo.return"(%0) : (tensor<f32>) -> ()
+  }, {
+  ^output_conversion(%arg0: tensor<f32>):
+    %0 = "stablehlo.convert"(%arg0): (tensor<f32>) -> (tensor<bf16>)
+    "stablehlo.return"(%0) : (tensor<bf16>) -> (tensor<bf16>)
+  }) {
+  dimensions = dense<1> : tensor<1xbf16>
+} : (tensor<1x6xbf16>, tensor<bf16>) -> tensor<1xbf16>
+```
+
+However, it is not clear how such operations will be lowered to other IR
+representations, like HLO, which does not support such additional computation
+blocks. IMO there is no additional benefit to support such conversion
+functions for regular type given that there already exists infrastructure
+(backend support, lowering passes) to support regular types w/o conversion
+functions. My proposal here would be to restrict the support to only quantized
+types.
+
+## Appendix
+
+To provide an estimate of specification changes needed to implement option #1
+I have attempted to provide the blueprint here.
+
+### Specification of reduce op
+
+```python
```

Comment:
Why are you wrapping the following sections with python code block?

Reply:
fixed.